docs/development/settings.js.md
# Configuring the dashboard - settings.js
settings.js defines a declarative configuration for our react-dash.

A simple example:

```eval_rst
.. literalinclude:: ../../examples/js_static_example/settings.js 
   :language: javascript
```

## Regions
### Boostrap rows
Regions are rendered into divs, allowing you to add bootstrap `row` class, and thus to organize the dashboard into rows using the responsive grid.

### Conditional Rendering
If `region.multi = true` then you can use conditional rendering to render the region. If this is the case, then the region should contain an `elements` object. The elements object has keys which each contain a different array of elements. The muli-region should also define a datahandler. The datahandler returns a string which is used as the key to chose an array from the elements object.

@@TODO add example


CHANGELOG.md
0.7.3 2017-12-13
----------------
- Add DataTables jQuery plugin for displaying tables

0.7.2 2017-10-23
----------------
- Pin React version to v.15.6.2 pending a v.2.0.0 release which should support React v.16.
- Update broken links to sample data in settings.js. Map & Chart example was linking to a data source in Dropbox. Update it to link to data sources within the repo.
- Add Keys to sample data to fix console.log errors.
- Add a new Link feature to the Highlight component.

0.7.1 2017-08-03
----------------
- Update the Highlight Component
- Fix default filters that were not being applied.
- Bugfix for failing routing.

0.6.x 2017-4-1
------
- Add cardClasses to components.
- Add regions to Card component:
  * header
  * subheader
  * topmatter
  * subheader2
  * topmatter2
  * footerHeader
  * footerSubheader
  * bottommatter
  * footerSubheader2
  * bottommatter2
Card regions can be set as props on the wrapped element, and as a consequence can be set via stateHandlers, allowing us to change the appearance of the card based on component state for wrapped elements.
- Add stateHandlers
State handlers fire on component mount and state update. They should be passed an `attr` attribute which tells the stateHandler which component prop to update. If the handler updates a cardVar (card region above or cardClasses narray, etc, the properties will be passed to the wrapping Card element for the wrapped component)
- Allow accordion regions to open/close with a click.
- Add accessible headers for card elements.
- Add accessible filter labels for the autocomplete component.
- Add support for props.dashWrapperClass in dashboard component
- Add support for props.isLink in highlight component

0.6.x 2017-3-7-2017 
-------------------
- Add stateHandlers
- Move Card wrapping to render method of 'Card-able' components. This allows
us to pass Card variables to card based on component state - so we can do
dynamic rendering based on conditional logic capture in stateHandlers
- add cardVariables to component state. These can be derived from props, or
can be generated computationally using stateHandlers. Also we could use this
mechanism in component code to pass component state or config  to the wrapping
Card

0.6.x 2017-2-27-2017 
--------------------
- Add Highlight component which provides a very basic profile type list

0.5.x 2017-2-23-2017
--------------------
- Dashboard.props.isFetching is set to true during dashboard instantiation and
should reset to false in Dashboard.getDashboardData() once all dashboard data
is loaded. 

0.5.x 2017-1-30-2017
--------------------
- Add Multicheckbox component - this works widd the onAction ->
AUTOCOMPLETE_CHANGE event. See /examples/settings.js for configuration
options.
- Filter Routing - EXPERIMENTAL - this can be disabed with
disableFilterRouting in settings

0.5.1 2017-1-10
---------------
- Add `npm run js-example` -- runs simple js config'd app from
/examples/js_static_example
- Add `npm run jsx-example` -- runs simple jsx flavor app from
/examples/jsx_static_example
- Documentation updates
- Add `disabledBy` prop to filters - allows other filters to disable a filter
when present as appliedFilter
- Chart component props.settings.tickFormat is d3 formatter string which is
turned into a function by the Chart component

0.5.0 2016-08-31
----------------
- Flux-ISH data flow in example app
Data is fetched, processed and updated at the App implementation level
(@@DASHBOARD LEVEL?). 
Only Dashboard level actions can change state.
Dashboard uses component settings from settings.js and dataHandlers (read:
reducers) to calculate component data. 
All components are passed static data at render.
For a component to update data / application state it must trigger an action
at the dashboard / app level
- Regions
Regions will render rows, allowing for easy responsive grid rendering.
- Multi Regions
Regions can now be define 'multi: true' in props, in which case the named
dataHandler will be called. The data handler should return a string which will
serve as a key, this key is used to select which sub-region to render from the
multi-region. Replaces Multi Component.
- data abstraction - dataKeys
We add an abstraction for dataKeys - a dataKey represents an individual
app-level data source, which contains query objects
- Filters
We have a common filter settings definition that includes the field to filter
and a willFilter array to determine which dataKey's data we are filtering...
see full documentation @@LINK 
Added BaseFilter component 
Move event handling to base class to unify filter implementations
Add example of app filtering in example app
- Added Checkbox Filter
- Removed Multi Component
- Add datahandlers to common library:
    - getXYBYQueryDataWhere
    - whiteListValues
    - inspect
- update datahandlers:
    - common.getXYBYQueryData 
    - add dashboardData.dataKey level of data abstraction (see
above)
    - common.groupByRange - remove check for isPieChart - use NVD3 handler instead
    - common.rekeySeries - return values array as array [output]

0.4.x 2016-08-31 
----------------
- Update data handling
- Add data table component
    - Rename Table component to DataTable to avoid namespace collisions
    - Add hideFilterHeader and hideControls props to component
    - ** NOTE ** DataTable filters rely on Dataset.query which is
    - ** NOTE ** provided by Backends (csv, google docs etc)
    - ** NOTE ** Will not-currently work with non-Datasrt dat
- Check for cardType in props before rendering element as Card
- Add multi component
- style prop on card elements rendered on card div - uses form { boderBottom:
 "1em" }
- Update Filter Handling for autocomplete component
- Move onAction to Dashboard component and include `field` value 
(defined in settings on Autocomplete component)
- Add appliedFilters to Dashboard state and pass to children in props
- Make appliedFiters avaiable in dataHandlers

0.3.7 2016-08-23
----------------

examples/settings.js
let stateIds =
      [
        { value: 1, label: 'AK' },
        { value: 30, label: 'NY' },
        { value: 2, label: 'AZ' },
        { value: 31, label: 'NC' },
        { value: 3, label: 'AR' },
        { value: 32, label: 'ND' },
        { value: 4, label: 'CA' },
        { value: 33, label: 'OH' },
        { value: 5, label: 'CO' },
        { value: 34, label: 'OK' },
        { value: 6, label: 'CT' },
        { value: 35, label: 'OR' },
        { value: 7, label: 'DE' },
        { value: 36, label: 'PA' },
        { value: 8, label: 'FL' },
        { value: 7, label: 'RI' },
        { value: 9, label: 'GA' },
        { value: 38, label: 'SC' },
        { value: 10, label: 'ID' },
        { value: 39, label: 'SD' },
        { value: 11, label: 'IL' },
        { value: 40, label: 'TN' },
        { value: 12, label: 'IA' },
        { value: 41, label: 'TX' },
        { value: 13, label: 'IA' },
        { value: 42, label: 'UT' },
        { value: 14, label: 'KS' },
        { value: 43, label: 'VT' },
        { value: 15, label: 'KT' },
        { value: 44, label: 'VA' },
        { value: 16, label: 'LA' },
        { value: 45, label: 'WA' },
        { value: 17, label: 'ME' },
        { value: 46, label: 'WV' },
        { value: 18, label: 'MD' },
        { value: 47, label: 'WI' },
        { value: 19, label: 'MA' },
        { value: 48, label: 'WY' },
        { value: 20, label: 'MI' },
        { value: 21, label: 'MN' },
        { value: 22, label: 'MI' },
        { value: 103, label: 'Central Region' },
        { value: 23, label: 'MO' },
        { value: 24, label: 'MT' },
        { value: 25, label: 'NE' },
        { value: 106, label: 'South Region' },
        { value: 26, label: 'NV' },
        { value: 27, label: 'NH' },
        { value: 28, label: 'NJ' },
        { value: 29, label: 'NM' },
        { value: 50, label: 'USA?'}
    ];
export var settings = {
  title: 'React-Dash v0.6.x -- Climate Indices by Year',
  dashWrapperClass: 'example-dash-wrapper',
  queries: {},
  doFilterRouting: false,
  // start dataResources example
  dataResources: {
    climateData: {
      fetchData: {
        type: 'backend',
        backend: 'csv',
        url: 'https://nucivic.github.io/react-dash/data/climate_indices.csv'
      },
    }
    // end dataResources example
  },

  regions: [
    {
      id: 'top-row filters-row',
      className: 'row',
      children: [
        {
          type: 'Autocomplete',
          className: 'col-md-6',
          name: 'specialty-autocomplete',
          key: 'specialty-autocomplete',
          className: 'specialty-autocomplete',
          field: 'YearMonth',
          action: 'filter', // sort / groupBy / etc
          willFilter: ['climateData'], // array of dkanDataResources keys that filters affect
          data: [
            [
              { label: '2010', value: '2010' },
              { label: '2011', value: '2011' },
              { label: '2012', value: '2012' },
              { label: '2013', value: '2013' },
              { label: '2014', value: '2014' },
              { label: '2015', value: '2015' },
            ]
          ],
          placeholder: 'Select year to filter dashboard...'
        }
      ]
    },
    {
      id: 'metrics-row',
      key: 'metrics-row',
      className: 'row',
      children: [
        {
          type: 'Metric',
          cardStyle: 'metric',
          key: 'metrics-row1',
          iconClass: 'fa fa-level-up',
          className: 'col-md-4',
          background: "#00b3b3",
          caption: 'Maximum Temp.',
          dataHandlers: [
            {
              name: 'getClimateMetric',
              field: 'TMAX'
            },
          ],
          stateHandlers: [
            {
              name: 'getMaxTempMetricColor',
              attr: 'bg'
            }
          ]
        },
        {
          type: 'Metric',
          cardStyle: 'metric',
          key: 'metrics-row2',
          iconClass: 'fa fa-level-down',
          className: 'col-md-4',
          background: '#009999',
          caption: 'Minimum Temp.',
          dataHandlers: [
            {
              name: 'getClimateMetric',
              field: 'TMIN'
            }
          ]
        },
        {
          type: 'Metric',
          cardStyle: 'metric',
          key: 'metrics-row3',
          iconClass: 'fa fa-fire',
          className: 'col-md-4',
          caption: 'Average Temp/',
          background: '#004d4d',
          dataHandlers: [
            {
              name: 'getClimateMetric',
              field: 'TAVG'
            }
          ]
        }
      ]
    },
    {
      id: 'map-row',
      className: 'row',
      children: [
        // choropleth
        {
          type: 'Choropleth',
          key: 'map-row-1',
          cardStyle: 'map',
          iconClass: 'fa fa-balance-scale',
          header: 'Palmer Hydrological Drought Index',
          format: 'topojson',
          dataHandlers: [
            {
              name: 'getMapData',
              stateArray: stateIds
            }
          ],
          dataKeyField: 'name',
          dataValueField: 'PHDI',
          geometryKeyField: 'name',
          geometry: 'https://nucivic.github.io/react-dash/data/map/usa.json', // topojson or geojson
          projection: 'albersUsa', // https://github.com/d3/d3/wiki/Geo-Projections
          scaleDenominator: .75,
          borderColor: 'yellow',
          noDataColor: 'yellow',
          topologyObject: 'usa',
          startColor: '#e6ffff',
          endColor: '#004d4d',
          dataClassification: 'equidistant',
          legend: {
            classesCount: 5,
            pallete:
              [
                "#e6ffff",
                "#00ffff",
                "#00cccc",
                "#00b3b3",
                "#009999",
                "#008080",
                "#004d4d"
              ]
          },
        },
      ]
    },
    {
      id: 'chart-row',
      className: 'row',
      children: [
        {
          type: 'Chart',
          key: 'chart-row-1',
          cardStyle: 'chart',
          header: 'Standard Precipitation Index',
          iconClass: 'fa fa-cloud',
          dataHandlers: [ 'getBarChartData' ],
          stateHandlers: [
            {
              name: 'isStatSignificant',
              attr: 'footer'
            }
          ],
          settings: {
            type: 'multiBarChart',
            x: 'x',
            y: 'y',
            height: 800
          }
        }
      ]
    },
    {
      id: 'fixed-data-table-row',
      className: 'row',
      children: [
        {
          type: 'DataTable',
          dataHandlers: [
            {
              name: 'getTableData'
            },
          ],
          key: 'fixed-data-table-row',
          header: 'Fixed Data Table 2',
          iconClass: 'fa fa-database',
          cardStyle: 'table',
          cardClasses: ['col-md-12', 'climate-data'],
          hideControls: true,
          hideFilterHeader: true,
          settings: {
            table: {
              rowHeight: 60,
              width: '100%',
              maxHeight: 300,
              headerHeight:60
            },
            columns: {
              flexGrow: 1,
              width: 150,
            },
          },
          overrides: {
            rows: {
              "max": {
                className: 'yelloW'
              }
            },
            columns: {
              "max": {
                className: 'greenCell'
              }
            },
            cells: {
              "min_3": {
                className: 'greenCell'
              }
            }
          }
        },
        
      ]
    },
    {
      id: 'datatable-row',
      className: 'row',
      children: [
        {
          type: 'ResponsiveDataTable',
          dataHandlers: [
            {
              name: 'getTableData'
            },
          ],
          key: 'datatables-row',
          header: 'Responsive DataTables',
          iconClass: 'fa fa-database',
          cardStyle: 'table',
          cardClasses: ['col-md-12', 'climate-data'],
          overrides: {
            rows: {
              "max": {
                className: 'yelloW'
              }
            },
            columns: {
              "max": {
                className: 'greenCell'
              }
            },
            cells: {
              "min_3": {
                className: 'greenCell'
              }
            }
          }
        },
        
      ]
    },    
    {
      id: 'highlight-row',
      className: 'row',
      children: [
        {
          type: 'Highlight',
          key: 'highlight1',
          header: 'Highlight',
          data: [
            {
              cols: [
                {rows:[
                  {label: 'label 1', val: 'Value 1'},
                  {label: 'Full Link', val: 'This is a full link', url:'http://google.com', isLink:true}
                ]},
                {rows:[
                  {label: 'label 2', val: 'Value 2'}
                ]}
              ]
            }
          ]
        }
      ]
    }        
  ]
};


let climateVars = {
			PCP: 'Precipitation Index',
			TAVG: 'Temperature Index',
			TMIN: 'Minimum Temperature Index',
			TMAX: 'Maximum Temperature Index',
			PDSI: 'Palmer Drought Severity Index',
			PHDI: 'Palmer Hydrological Drought Index',
			ZNDX: 'Palmer Z-Index',
			PMDI: 'Modified Palmer Drought Severity Index',
			CDD: 'Cooling Degree Days',
			HDD: 'Heating Degree Days',
			SPnn: 'Standard Precipitation Index'
		};
